### PR TITLE
Update Network classes

### PR DESCRIPTION
**Pull request by mcwimm**
Update basic network concepts for bg and growth:
- Added different variants (V0-V2) to describe root graft formation and resource allocation during this process
- Added input parameters

### CODE REVIEW

File: Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Variant used to simulate root graft formation (RGF). 
+
+Possible inputs are: "V0", "V1" or "V2"
+
+Meaning: 
+
+- 'V0' ... RGF does not effect tree growth and trees are immediately grafted if roots are in contact.
+- 'V1' ... The duration of RGF is fixed (2 years) and tree growth is reduced by a factor f_growth.
+- 'V2' ... The duration of RGF is determined by a minimum radius of the grafted roots which in turn is defined by the factor f_radius. The  energy for root graft growth is taken from the energy designated to radial growth: w_gr = f_growth ∗ w_rstem.
```
Comment (jbathmann) [inline Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md:9]: Have you got more instructive names for this?
eg.: V0 -> RGFWithoutEffect
V1 -> RGFFixedTime

Whatever - you might have more creativity on this..

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+
+
+optional: Network-parameter which should be added to the output. You can define add an arbitrary number of parameters.
```
Comment (jbathmann) [inline Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md:3]: Do you need to add it here, if you want to get it for the output?

File: TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py [hunk 3]
```diff
@@ -25,6 +25,8 @@ def __init__(self, args):
             self.iniFixedSalinity(args)
         elif case == "SimpleNetwork":
             self.iniSimpleNetwork(args)
+        elif case == "SimpleNetworkVar":
+            self.iniSimpleNetworkVar(args)
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py:29]: I thought you don't need this anymore?

File: TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py [hunk 4]
```diff
@@ -58,6 +60,10 @@ def iniSimpleNetwork(self, args):
         from .SimpleNetwork import SimpleNetwork
         self.concept = SimpleNetwork(args)
 
+    def iniSimpleNetworkVar(self, args):
+        from .SimpleNetworkVar import SimpleNetworkVar
+        self.concept = SimpleNetworkVar(args)
+
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py:66]: s.o.

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py [hunk 5]
```diff
@@ -118,7 +139,7 @@ def addTree(self, tree):
 
         self._psi_leaf.append(parameter["leaf_water_potential"])
         self._psi_height.append(
-            (2 * geometry["r_crown"] + geometry["h_stem"]) * 9810)
+            -(2 * geometry["r_crown"] + geometry["h_stem"]) * 9810)
         self._psi_top = np.array(self._psi_leaf) - np.array(self._psi_height)
         self._psi_osmo = np.array(
             [0] * self.n_trees)  # Salinity is 0 ppt is the basic scenario
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:145]: can you now calculate
psi_ges = sum_i psi_i or do you have to be careful with the signs?

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py [hunk 6]
```diff
@@ -25,33 +27,142 @@ def progressTree(self, tree, aboveground_resources, belowground_resources):
...
+            SimpleBettina.treeGrowthWeights(self)
+
     def growthResources(self):
-        # Simple bettina get growth resources
         SimpleBettina.growthResources(self)
-        self.rootGraftFormation()
+        if self.variant == "V0":
+            self.rootGraftFormationV0()
+        if self.variant == "V1":
+            self.growthResourcesV1()
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:71]: This is confusing?
Why do you either do a graft Formation or calculate resources?

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py [hunk 7]
```diff
@@ -25,33 +27,142 @@ def progressTree(self, tree, aboveground_resources, belowground_resources):
...
+    # defined in SimpleBettina. In addition, the function calls the root
+    # graft formation function, if the tree is currently in the process of
+    # root graft formation.
+    def treeGrowthWeightsV2(self):
+        # Simple bettina get growth weigths
+        SimpleBettina.treeGrowthWeights(self)
+
+        # If self.r_gr_min exists the tree is currently in the process of rgf
+        if self.r_gr_min:
+            self.rootGraftFormationV2()
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:86]: an here it is a graft formation again...

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py [hunk 8]
```diff
@@ -25,33 +27,142 @@ def progressTree(self, tree, aboveground_resources, belowground_resources):
...
-    # It is assumed that the process will take 2 years (this is subject to change).
-    def rootGraftFormation(self):
+    # It is assumed that the process will take 2 years (this is subject to
+    # change).
+    def rootGraftFormationV1(self):
         if self.rgf != -1:
-            self.grow = self.grow / 2
+            self.grow = self.grow * (1 - self.f_growth)
             self.rgf = self.rgf + 1
             if round(self.rgf * self.time / 3600 / 24 / 365, 3) >= 2:
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:112]: I am not sure, but where do you get "time" from? Is it independent of the startpoint?

File: test/mcw/se_script.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
```
Comment (jbathmann) [inline test/mcw/se_script.py:1]: does this belong here?

**Review by jbathmann — COMMENTED** [review]
This PR is pretty big. It is hard to follow. Please provide smaller chunks in the next PR's. ;-)

File: TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py (hunk 3 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py:29]: right. solved

File: TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py (hunk 4 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/BelowgroundCompetition.py:66]: solved

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py (hunk 5 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:145]: I don't get this. But the '-' sign is important (and now correct)

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py (hunk 6 above)
Comment (mcwimm) [reply TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:71]: Because the different concepts alter tree growth at different points in the model, and 'treeGrowthWeights' and 'growthResources' are the basic BETTINA functions called by the main 'progressTree' function. So I thought to calls rootGraftFormation variants from these functions. Therefore, I don't need to re-write/ copy BETTINA functions

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py (hunk 7 above)
Comment (mcwimm) [reply TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:86]: s.o.

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py (hunk 8 above)
Comment (mcwimm) [reply TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:112]: This is the time step length, which is calculated in the parent class (SimpleBettina) in the method 'prepareNextTimeStep'

File: test/mcw/se_script.py (hunk 9 above)
Comment (mcwimm) [reply test/mcw/se_script.py:1]: Don't know. Where can I store such scripts?

File: Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md (hunk 1 above)
Comment (mcwimm) [reply Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md:9]: I'll think about it...

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md (hunk 2 above)
Comment (mcwimm) [reply Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md]: I think so. It's the same as for growth or geometry output.
However, I completed the text to match the other descriptions.

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+
+
+optional: Network-parameter which should be added to the output. 
+You can define an arbitrary number of parameters as long as they are available for the network concepts. 
+To check availability of network information, please see network concept tag documentation.
```
Comment (jbathmann) [inline Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md:5]: Sorry, but can you update this in the other outputs?

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py (hunk 5 above)
Comment (jbathmann) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:145]: ok

Comment (jbathmann) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:145]: is psi_ges = psi_top + psi_height .. or ist ist  psi_ges = psi_top - psi_heigh...?

File: TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py (hunk 8 above)
Comment (jbathmann) [reply TreeModelLib/GrowthAndDeathDynamics/NetworkBettina/NetworkBettina.py:112]: ok

File: test/mcw/se_script.py (hunk 9 above)
Comment (jbathmann) [reply test/mcw/se_script.py:1]: on your local device?

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py (hunk 5 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:145]: it's: psi_ges = psi_leaf - psi_height - psi_osmo

File: test/mcw/se_script.py (hunk 9 above)
Comment (mcwimm) [reply test/mcw/se_script.py]: ok. removed

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md (hunk 10 above)
Comment (mcwimm) [reply Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFile/s_network_output.md:5]: ok. done

File: Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md (hunk 1 above)
Comment (mcwimm) [reply Documentation/ProjectFile/MangaProject/tree_dynamics/tree_growth_and_death/NetworkBettina/s_variant.md]: Renamed to
- V0_instant  
- V1_fixed
- V2_adapted 
... with the suffix describing the duration of root graft formation

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFilePerGroup/s_network_output.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+
+
+optional: Network-parameter which should be added to the output. 
+You can define an arbitrary number of parameters as long as they are available for the network concepts. 
+To check availability of network information, please see network concept tag documentation.
```
Comment (jbathmann) [inline Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFilePerGroup/s_network_output.md:5]: Have you checked whether it works?

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py [hunk 12]
```diff
@@ -196,8 +251,43 @@ def updatePartnerIdices(self):
...
+                    else:
+                        # trees are only put in the dict. if the occur more than
+                        # ones, i.e. both partners have finished rgf
+                        link_list.append({vertex, neighbour})
+                        self.setKeyDictionary(self.graph_dict, vertex, neighbour)
+
+    ## This function creates a graph dictionary from the partner indices.
+    # Includes all links, i.e. also if not both partners finished rgf
+    # NOT USED
+    def makeGraphDictionaryAll(self):
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:290]: hmm
Are you using this function? if not - why does it appear here?

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py [hunk 13]
```diff
@@ -196,8 +251,43 @@ def updatePartnerIdices(self):
...
+    ## Function that adds keys and values to dictionary
+    def setKeyDictionary(self, dictionary, key, value):
+        if key not in dictionary:
+            dictionary[key] = value
+        elif len(dictionary[key]) == 0:
+            dictionary[key] = {value}
+        else:
+            dictionary[key] = dictionary[key] | {value}
+
+    ## Funktion that makes a graph dictionary only with functional grafts,
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py:263]:
```suggestion
    ## Function that makes a graph dictionary only with functional grafts,
```

**Comment by jbathmann** [comment]
Hard to impossible to follow. But anyways - if it works, it works. ;-)
Do you have test cases for your system which we could use as benchmarks?

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py (hunk 13 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py]: resolved

File: TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py (hunk 12 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SimpleNetwork/SimpleNetwork.py]: resolved

File: Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFilePerGroup/s_network_output.md (hunk 11 above)
Comment (mcwimm) [reply Documentation/ProjectFile/MangaProject/tree_output/OneTimestepOneFilePerGroup/s_network_output.md:5]: yes, it works

**Comment by mcwimm** [comment]
> Hard to impossible to follow. But anyways - if it works, it works. ;-)
> Do you have test cases for your system which we could use as benchmarks?

I added a test case to SmallTests - but I don't get how you validate the results

**Comment by jbathmann** [comment]
> > Hard to impossible to follow. But anyways - if it works, it works. ;-)
> > Do you have test cases for your system which we could use as benchmarks?
> 
> I added a test case to SmallTests - but I don't get how you validate the results

Thanks. Currently, we can not validate the results. @jvollhueter is planning a test against previous results, such that we can see if anything changes significantly, with a new PR.

However, there is a testscript stored in /test/SmallTests, which should automatically start your test. I checked it and it seems to work.
